src/main/java/com/exactpro/th2/healer/grpc/HealerServiceImpl.java

```diff
@@ -0,0 +1,196 @@
+import com.exactpro.th2.crawler.dataservice.grpc.CrawlerInfo;
+import com.exactpro.th2.crawler.dataservice.grpc.DataServiceGrpc;
+import com.exactpro.th2.crawler.dataservice.grpc.DataServiceInfo;
+import com.exactpro.th2.crawler.dataservice.grpc.EventDataRequest;
+import com.exactpro.th2.crawler.dataservice.grpc.EventResponse;
+import com.exactpro.th2.crawler.dataservice.grpc.Status;
+import com.exactpro.th2.dataprovider.grpc.EventData;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.exactpro.th2.common.message.MessageUtils.toJson;
+
+
+public class HealerServiceImpl extends DataServiceGrpc.DataServiceImplBase {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HealerServiceImpl.class);
+
+    private final HealerConfiguration configuration;
+    private final CradleStorage storage;
+    private final Map<String, InnerEvent> cache;
+    private final ConcurrentHashMap.KeySetView<CrawlerId, Boolean> knownCrawlers;
+
+    public HealerServiceImpl(HealerConfiguration configuration, CradleStorage storage) {
+        this.configuration = configuration;
+        this.storage = Objects.requireNonNull(storage, "Cradle storage cannot be null");
+        this.cache = new EventsCache<>(configuration.getMaxCacheCapacity());
+        this.knownCrawlers = ConcurrentHashMap.newKeySet();
+    }
+
+    @Override
+    public void crawlerConnect(CrawlerInfo request, StreamObserver<DataServiceInfo> responseObserver) {
+        try {
+
+            if (LOGGER.isInfoEnabled()) {
+                LOGGER.info("crawlerConnect request: {}", toJson(request, true));
+            }
+
+
+            knownCrawlers.add(request.getId());
+
+            DataServiceInfo response = DataServiceInfo.newBuilder()
+                    .setName(configuration.getName())
+                    .setVersion(configuration.getVersion())
+                    .build();
+
+            LOGGER.info("crawlerConnect response: {}", response);
+
+            responseObserver.onNext(response);
+            responseObserver.onCompleted();
+        } catch (Exception e) {
+            responseObserver.onError(e);
+            LOGGER.error("crawlerConnect error: " + e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public void sendEvent(EventDataRequest request, StreamObserver<EventResponse> responseObserver) {
+        try {
+            LOGGER.info("sendEvent request: {}", request);
+
+            if (!knownCrawlers.contains(request.getId())) {
+
+                if (LOGGER.isWarnEnabled()) {
+                    LOGGER.warn("Received request from unknown crawler with id {}. Sending response with HandshakeRequired = true", toJson(request.getId()));
```

Comment:
The method `toJson` without a boolean parameter is deprecated as far as I remember. Please, check it and replace the deprecated method